### PR TITLE
esphome 2023.11.2

### LINKED ISSUE #153593: Module named 'idna' missing in esphome 2023.10.6
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
HOMEBREW_VERSION: 4.1.19
ORIGIN: https://github.com/Homebrew/brew
HEAD: 7f745d04d90eaec76c8d5830d6ce6a0d7db888b3
Last commit: 2 days ago
Core tap JSON: 07 Nov 20:42 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.10 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Applications/Xcode.app/Contents/Developer/usr/bin/git
Curl: 8.1.2 => /usr/bin/curl
macOS: 13.6-arm64
CLT: 15.0.0.0.1.1694021235
Xcode: 15.0.1
Rosetta 2: false
```

### Verification

- [X] My "`brew doctor` output" says `Your system is ready to brew.` and am still able to reproduce my issue.
- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.
- [X] I searched for recent similar issues at https://github.com/Homebrew/homebrew-core/issues?q=is%3Aissue and found no duplicates.

### What were you trying to do (and why)?

I can't run esphome and get the following error when : ModuleNotFoundError: No module named 'idna'

https://github.com/Homebrew/homebrew-core/commits/e2bba732d6a3f5289c7357ce5a3ffa8c909aa61a/Formula/e/esphome.rb

esphome/2023.10.6

### What happened (include all command output)?

Traceback (most recent call last):
File "/opt/homebrew/bin/esphome", line 8, in 
sys.exit(main())
^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/main.py", line 1036, in main
return run_esphome(sys.argv)
^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/main.py", line 1023, in run_esphome
rc = POST_CONFIG_ACTIONS[args.command](args, config)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/main.py", line 409, in command_compile
exit_code = compile_program(args, config)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/main.py", line 216, in compile_program
rc = platformio_api.run_compile(config, CORE.verbose)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/platformio_api.py", line 108, in run_compile
return run_platformio_cli_run(config, verbose, *args)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/platformio_api.py", line 101, in run_platformio_cli_run
return run_platformio_cli(*command, **kwargs)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/esphome/platformio_api.py", line 90, in run_platformio_cli
import platformio.main
File "/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/platformio/main.py", line 21, in 
from platformio import version, exception, maintenance
File "/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/platformio/maintenance.py", line 22, in 
from platformio import version, app, exception, fs, telemetry
File "/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/platformio/telemetry.py", line 25, in 
import requests
File "/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/requests/init.py", line 147, in 
from . import packages, utils
File "/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/requests/packages.py", line 16, in 
locals()[package] = import(package)
^^^^^^^^^^^^^^^^^^^
ModuleNotFoundError: No module named 'idna'

### What did you expect to happen?

esphome compile
without errors

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
brew install esphome
brew install platformio
esphome compile esp32-C3.yaml
```

### PR DESCRIPTION
**Pull request by cho-m**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [ ] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [ ] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

Starting with trying to add test to make sure failure seen.

Will work on quick pth fix afterward.

Fixes #153593

### CODE REVIEW

**Comment by chenrui333** [comment]
there is a new release now, `2023.11.0`

**Comment by p-linnane** [comment]
We can vendor `python-argcomplete` here.

File: Formula/e/esphome.rb [hunk 1]
```diff
@@ -110,15 +115,15 @@ class Esphome < Formula
...
+    url "https://files.pythonhosted.org/packages/db/47/85eafb277f6ef78e1a1895cc72f0035dfa6a5e51396134eb9ce21564c72f/zeroconf-0.123.0.tar.gz"
+    sha256 "c50f24c9a7a6c7ba4bb301defad03a9d85dffb3ee280953de44eee9f432f2550"
   end
 
   def install
     virtualenv_install_with_resources
 
     site_packages = Language::Python.site_packages("python3.12")
-    paths = %w[platformio].map { |p| Formula[p].opt_libexec/site_packages }
+    paths = %w[platformio uvicorn].map { |p| Formula[p].opt_libexec/site_packages }
```
Comment (singingwolfboy) [inline Formula/e/esphome.rb:126]: Don't you need to add `depends_on "uvicorn"` for this to work?

Comment (cho-m) [reply Formula/e/esphome.rb:126]: PTH files are not recursively loaded. See failure in #153593.

Comment (cho-m) [reply Formula/e/esphome.rb:126]: Maybe a better option would be to use `site.addsitedir` similar to how Homebrew automatically does https://github.com/Homebrew/brew/blob/master/Library/Homebrew/language/python.rb#L162

That will get Python to load the PTH files from the specified directory, e.g.

- Existing formula:
    ```
    ❯ /opt/homebrew/opt/esphome/libexec/bin/python3.12 -I -v -c 'exit()' 2>&1 | rg 'Processing'
    Processing global site-packages
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew-deps.pth'
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew_deps.pth'
    Processing .pth file: '/opt/homebrew/opt/python-setuptools/lib/python3.12/site-packages/distutils-precedence.pth'
    Processing .pth file: '/opt/homebrew/opt/protobuf/lib/python3.12/site-packages/protobuf-4.25.0-py3.12-nspkg.pth'
    Processing user site-packages
    Processing global site-packages
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew-deps.pth'
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew_deps.pth'
    Processing .pth file: '/opt/homebrew/opt/python-setuptools/lib/python3.12/site-packages/distutils-precedence.pth'
    Processing .pth file: '/opt/homebrew/opt/protobuf/lib/python3.12/site-packages/protobuf-4.25.0-py3.12-nspkg.pth'
    Processing .pth file: '/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/distutils-precedence.pth'
    Processing .pth file: '/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/homebrew-gpgme-1.23.1.pth'
    Processing .pth file: '/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/protobuf-4.25.0-py3.12-nspkg.pth'
    ```
- Using `site.addsitedir`:
    ```
    ❯ /opt/homebrew/opt/esphome/libexec/bin/python3.12 -I -v -c 'exit()' 2>&1 | rg 'Processing'
    Processing global site-packages
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew-deps.pth'
    Processing .pth file: '/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/homebrew-deps.pth'
    Processing .pth file: '/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/homebrew_deps.pth'
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew_deps.pth'
    Processing .pth file: '/opt/homebrew/opt/python-setuptools/lib/python3.12/site-packages/distutils-precedence.pth'
    Processing .pth file: '/opt/homebrew/opt/protobuf/lib/python3.12/site-packages/protobuf-4.25.0-py3.12-nspkg.pth'
    Processing user site-packages
    Processing global site-packages
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew-deps.pth'
    Processing .pth file: '/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/homebrew-deps.pth'
    Processing .pth file: '/opt/homebrew/opt/platformio/libexec/lib/python3.12/site-packages/homebrew_deps.pth'
    Processing .pth file: '/opt/homebrew/Cellar/esphome/2023.10.6/libexec/lib/python3.12/site-packages/homebrew_deps.pth'
    Processing .pth file: '/opt/homebrew/opt/python-setuptools/lib/python3.12/site-packages/distutils-precedence.pth'
    Processing .pth file: '/opt/homebrew/opt/protobuf/lib/python3.12/site-packages/protobuf-4.25.0-py3.12-nspkg.pth'
    Processing .pth file: '/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/distutils-precedence.pth'
    Processing .pth file: '/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/homebrew-gpgme-1.23.1.pth'
    Processing .pth file: '/opt/homebrew/Cellar/python@3.12/3.12.0/Frameworks/Python.framework/Versions/3.12/lib/python3.12/site-packages/protobuf-4.25.0-py3.12-nspkg.pth'
    ```

**Review by chenrui333 — APPROVED** [review]

**Pull request merged**